### DISCUSSION TITLE
Create orgue file

### DISCUSSION THREAD

**Discussion by larspalo** [GrandOrgue/grandorgue#330] [General]
Originally posted 2020-06-24 15:45:26.619000 by tbpr in GrandOrgue discussion forum on SourceForge<br><br>I would like to create a .orgue file for one of my sample sets. I&#039;ve looked at the documentation, but I need a little more information about the process.

I understand the samples need to be compressed using wavpack. Does the version matter?

When I compress the .wav files using wavpack, by default it creates a new file with the .wv extension. Is this what the .orgue format is expecting?

There are potentially 10s of thousands of samples in a sample set. After compressing they then need to to be added to the zip (.orgue) file. Is it possible to use a script to assemble all of the files from many different directories into the .orgue file? 

Has anyone created a .orgue file for a large sample set? Any hints or ideas how to procede?

**Comment by larspalo** [comment]
Originally posted 2020-07-10 08:46:08.019000 by baskb in GrandOrgue discussion forum on SourceForge<br><br>Not for a large sampleset but for my sampleset of the Onderhorst cabinet organ and I did not apply compression to the samples . In my opinion, the principle remains the same. Pack the set of files with a Zip program (I used 7Zip, Compression Level=0) and rename &quot;.zip&quot; to &quot;.orgue&quot;.

**Comment by larspalo** [comment]
Originally posted 2020-07-22 23:49:13.935000 by [Martin Koegler](https://github.com/e9925248) in GrandOrgue discussion forum on SourceForge<br><br>Look at my post about GrandOrgueTool. It will help you with packing a working on-disk organ into an orgue file.

**Reply by jason-barba** [reply]
Lifting this thread.

I have searched for info on the GrandOrgueTool and the tool itself , but no success. Where can I find?

Also, is there a description of the organindex.ini? I notice the [General] and [Organ999] paragraphs (or objects).

//Jens

**Reply by larspalo** [reply]
GrandOrgueTool is built together with GrandOrgue and is available in the same bin folder as GrandOrgue. If you run it with

`GrandOrgueTool -h`

you get a short help description.

It's supposed to work on an already existing organ that you've opened/loaded in GO and package it into an .orgue package. It seems like it doesn't do the wavpack compression of the samples, though. For this I usually run a small shell script in the root of the sampleset:
~~~
for a in `find -name \*.wav`; do wavpack -h -x6 $a -o $a.new && mv $a.new $a; done
~~~

For using GrandOrgueTool to create a package you have to specify at least the following necessary parameters:

`GrandOrgueTool -c -o OrganPackageFilename -i  SamplesetRootDirectory -t OrganPackageTitle -o1 OdfName`

additionally, multiple odfs can be specified with -o1 to -o10 and if your organ package does depend on other organ packages you can specify them with -p1 to -p5. Of course, you should replace my dummy explanatory names with actual relevant ones.

PS, if you really don't want to use GrandOrgueTool it's of course possible to do everything manually too.

**Reply by jason-barba** [reply]
Thanks a plenty, Lars.

**Reply by jason-barba** [reply]
Follow up Qs: I suppose GrandOrgue can play out wavpacked wav files, right? Otherwise there have to be an unpack function when Loading the .orgue file. And, what happens if the wav-files that are packed by GrandOurgueTool already are wavpack compressed when applying wavpack on them?

**Reply by larspalo** [reply]
Yes, GO can handle wavpacked files. (It seems that they are silently unpacked at load time when the samples are read into memory anyway)

At least in the version of wavpack that I have, if the files already are wavpacked, you'll be asked a question if you want to overwrite the file (yes/no/all). Doing so seems to just recreate the same wavpacked file without any issues, but YMMV.

As a side note, the GO help states that only version 4 of wavpack is supported but version 5 seems to be fully backwards compatible so it should be possible to use with GO too.

**Reply by jason-barba** [reply]
Thanks!

Here comes another Q: Do all .organ files in a folder have to be included (o1, o2, o3 ...) in the GrandOurgueTool command for producing av working .orgue file? In my test there where three .organ files but I only entered the one I was interested in in the GrandOrgueToll command. All three where included in the .orgue file evenso. But after Installing I only got loads of error messages when  Load. Files, mainly wav) where not found.

**Reply by larspalo** [reply]
Ok, I'll correct myself slightly.

GrandOrgueTool does wavpack the samples, so that step is actually not necessary if you're using the tool.

All files in the sampleset directory is included in the .orgue package, but they are not necessarily used. Thus, you need to specify all the odfs you want to be able to use with -o1 -o2 ... and so on.

The reason you're getting files not found is because the file paths include .\ after the = sign. Ie. for the organ package to work all paths in the odf(s) must be specified like:

`Pipe001=Subbas16\036-C.wav`

This includes all file paths, not only the samples.

**Reply by jason-barba** [reply]
Aha! Does GO accept both \ and / as separators in paths?

**Reply by larspalo** [reply]
It kind of works with both, but the canonical separator is \ so to avoid any warnings the backslash \ character should be used and nothing else.

**Comment by jason-barba** [comment]
Finally I managed :-)
Thanks for helping out, Lars.

//Jens

**Reply by larspalo** [reply]
No problem, I've documented the process in the wiki at [Organ package creation (GrandOrgueTool)](https://github.com/GrandOrgue/grandorgue/wiki/Organ-package-creation-(GrandOrgueTool)).

**Reply by jason-barba** [reply]
Missed that link ;-)

**Comment by vpoguru** [comment]
The GrandOrgueTool currently does not run on Apple Silicon Macs, because brew does not install wavpack in the folder /usr/local/opt/wavpack, which seems to be hardcoded in the tool, but in /opt/homebrew/opt/wavpack.

Is there any way to use the GrandOrgueTool anyway?

**Reply by larspalo** [reply]
Can't say I've ever seen any hardcoded paths to wavpack in the sources. (BTW, on my Ubuntu computer wavpack is in /usr/bin/wavpack) It's just found with pkg_check_modules (pkgconfig) and included as normal with

`#include <wavpack/wavpack.h>`

If GrandOrgueTool cannot use wavpack on your system can GrandOrgue use it (like opening the PiteaMHS .orgue package)?

**Comment by vpoguru** [comment]
I was able to run the PiteaMHS.orgue package. I am also able to  run wavpack (installed with Homebrew) from the command line.

When I run the GrandOrgueTool with the -h option, I get the following error message:

    dyld[67698]: Library not loaded: /usr/local/opt/wavpack/lib/libwavpack.1.dylib
      Referenced from: /Applications/GrandOrgue.app/Contents/MacOS/GrandOrgueTool
      Reason: tried: '/usr/local/opt/wavpack/lib/libwavpack.1.dylib' (no such file),
      '/usr/local/lib/libwavpack.1.dylib' (no such file), '/usr/lib/libwavpack.1.dylib' (no such file)
    zsh: abort      ./GrandOrgueTool -h

When I then manually copy the file libwavpack.1.dylib to the folder specified above, I get the following error message:

    dyld[77790]: Library not loaded: /usr/local/opt/wavpack/lib/libwavpack.1.dylib
      Referenced from: /Applications/GrandOrgue.app/Contents/MacOS/GrandOrgueTool
      Reason: tried: '/usr/local/opt/wavpack/lib/libwavpack.1.dylib' (no such file),
      '/usr/local/lib/libwavpack.1.dylib'
      (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')),
      '/usr/lib/libwavpack.1.dylib' (no such file)
    zsh: abort      ./GrandOrgueTool -h

I suppose I need to compile GrandOrgue for Apple Silicon, but so far it has always failed.

**Reply by larspalo** [reply]
I have close to zero experience with any Apple products... But the error message seems to suggest a problem with architecture. Maybe @oleg68 knows more?

**Reply by oleg68** [reply]
Unfornunally, I don't have any.

**Comment by vpoguru** [comment]
I was now able to successfully build the app on macOS. However, it crashes immediately. Attached you will find the crash log.

[GrandOrgueCrashLog.zip](https://github.com/GrandOrgue/grandorgue/files/7780613/GrandOrgueCrashLog.zip)

... and the GrandOrgueTool still does not run.

**Comment by oleg68** [comment]
@vpoguru the error is 

```
Exception Type:        EXC_BAD_ACCESS (SIGKILL (Code Signature Invalid))
```

and the stack trace does not contain any GrandOrgue code.

```
0   dyld                          	       0x102d1e2c8 dyld3::MachOFile::compatibleSlice(Diagnostics&, void const*, unsigned long, char const*, dyld3::Platform, bool, dyld3::GradedArchs const&) + 76
1   dyld                          	       0x102d02758 invocation function for block in dyld4::JustInTimeLoader::makeJustInTimeLoaderDisk(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&, bool, unsigned int) + 96
2   dyld                          	       0x102d02758 invocation function for block in dyld4::JustInTimeLoader::makeJustInTimeLoaderDisk(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&, bool, unsigned int) + 96
3   dyld                          	       0x102d07b44 dyld4::SyscallDelegate::withReadOnlyMappedFile(Diagnostics&, char const*, bool, void (void const*, unsigned long, bool, dyld4::FileID const&, char const*) block_pointer) const + 128
4   dyld                          	       0x102d026c4 dyld4::JustInTimeLoader::makeJustInTimeLoaderDisk(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&, bool, unsigned int) + 204
5   dyld                          	       0x102cfc580 invocation function for block in dyld4::Loader::getLoader(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&) + 1412
6   dyld                          	       0x102cfba70 dyld4::Loader::forEachResolvedAtPathVar(dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&, dyld4::ProcessConfig::PathOverrides::Type, bool&, void (char const*, dyld4::ProcessConfig::PathOverrides::Type, bool&) block_pointer) + 780
7   dyld                          	       0x102cfb6d8 invocation function for block in dyld4::Loader::forEachPath(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&, void (char const*, dyld4::ProcessConfig::PathOverrides::Type, bool&) block_pointer) + 116
8   dyld                          	       0x102cf1e90 dyld4::ProcessConfig::PathOverrides::forEachPathVariant(char const*, dyld3::Platform, bool, bool&, void (char const*, dyld4::ProcessConfig::PathOverrides::Type, bool&) block_pointer) const + 672
9   dyld                          	       0x102cfb64c dyld4::Loader::forEachPath(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&, void (char const*, dyld4::ProcessConfig::PathOverrides::Type, bool&) block_pointer) + 172
10  dyld                          	       0x102cfbdf8 dyld4::Loader::getLoader(Diagnostics&, dyld4::RuntimeState&, char const*, dyld4::Loader::LoadOptions const&) + 840
11  dyld                          	       0x102d00968 invocation function for block in dyld4::JustInTimeLoader::loadDependents(Diagnostics&, dyld4::RuntimeState&, dyld4::Loader::LoadOptions const&) + 380
12  dyld                          	       0x102d1cffc invocation function for block in dyld3::MachOFile::forEachDependentDylib(void (char const*, bool, bool, bool, unsigned int, unsigned int, bool&) block_pointer) const + 148
13  dyld                          	       0x102ce9f98 dyld3::MachOFile::forEachLoadCommand(Diagnostics&, void (load_command const*, bool&) block_pointer) const + 168
14  dyld                          	       0x102d1ce3c dyld3::MachOFile::forEachDependentDylib(void (char const*, bool, bool, bool, unsigned int, unsigned int, bool&) block_pointer) const + 172
15  dyld                          	       0x102d006d0 dyld4::JustInTimeLoader::loadDependents(Diagnostics&, dyld4::RuntimeState&, dyld4::Loader::LoadOptions const&) + 164
16  dyld                          	       0x102ced618 dyld4::prepare(dyld4::APIs&, dyld3::MachOAnalyzer const*) + 1076
17  dyld                          	       0x102ced0d4 start + 488
```

So the reason is wrong build.

**Comment by vpoguru** [comment]
Attached you will find the redirected output of the complete build process.

[build.zip](https://github.com/GrandOrgue/grandorgue/files/7782285/build.zip)

**Comment by oleg68** [comment]
@vpoguru unfortunally, I don't know what is the right build process for Apple M1.

**Comment by vpoguru** [comment]
I am one step further. The GrandOrgueTool (compiled on Apple Silicon) now runs on Apple Silicon. The program seems to be even self-containing, as it can be moved out of the application package.

I am quite impressed how effective this package manager is. The resulting .orgue file is even significantly smaller than a .zip or .rar archive.

I am struggling a bit with non-ascii symbols. Regardless of whether the ODF and organindex.ini are encoded with the Windows 1252 or UTF-8 character set, non-ascii symbols get messed up.

Any chance to implement a progress indicator? At the beginning I thought that the program hangs until I found out that it takes several hours to create an .orgue file.

**Reply by jason-barba** [reply]
> Any chance to implement a progress indicator? At the beginning I thought that the program hangs until I found out that it takes several hours to create an .orgue file.

+1 on that. I opened another shell and checked the progress with ls -l ;-)

**Comment by paolo2504** [comment]
Looking at this post i ask if it is possible for developers to check why grandorguetool doesn't work under Windows (10 64bit).
thanks.

**Reply by oleg68** [reply]
Please create an issue about it.

Sorry, I see #949. It is already here.

**Comment by paolo2504** [comment]
Done.
